src/lsp4clj/server.clj

```diff
@@ -199,21 +256,16 @@
                        join]
   protocols.endpoint/IEndpoint
   (start [this context]
-    (let [;; a thread so language server can use >!! and so that receive-message
-          ;; can use (>!! output-ch) to respect back pressure from clients that
-          ;; are slow to read.
-          pipeline (async/thread
-                     (loop []
-                       (if-let [message (async/<!! input-ch)]
-                         (do
-                           (receive-message this context message)
-                           (recur))
-                         (async/close! output-ch))))]
+    ;; Start receiving messages.
+    (let [pipeline (run-pipeline this context input-ch)]
+      ;; Wait to stop receiving messages.
       (async/go
-        ;; Wait for pipeline to close. This indicates input-ch was closed and
-        ;; that now output-ch is closed.
+        ;; When pipeline closes, it indicates input-ch has closed. We're done
+        ;; receiving.
         (async/<! pipeline)
-        ;; Do additional cleanup.
+        ;; Do cleanup.
+        ;; TODO: do we really know that we've finished putting to output-ch?
+        (async/close! output-ch)
```

Comment:
The [docs](https://clojuredocs.org/clojure.core.async/close!) for `close!` say A) "The channel will no longer accept any puts", B) "Data in the channel remains available for taking", and C) "Logically closing happens after all puts have been delivered."

At this point the input-ch has been closed, which means any messages that were read before the channel was closed have been put on the channel (C). However, the takes off of it, the takes which then forward the messages to the language server, may or may not have happened (B). And even if the language server has received some messages, if it responds after this line closes the output-ch, the responses will be dropped (A).

All that to say, I think it's possible for the lsp4clj server to drop the language server's final few responses.

It doesn't really matter though, because neither clojure-lsp nor Scarlet call `shutdown` on the lsp4clj server until they've received the `exit` notification, which is the client indicating it no longer expects any responses anyway.